### PR TITLE
feat(storage) : add transient storage for cancun

### LINKED ISSUE #260: Transient storage for contracts
### Describe the feature

Support transient storage introduced in Cancun, `TSTORE (0x5D)` and `TLOAD (0x5D)`

### Solution

1. need to introduce CANCUN opcodes in [opcodes](https://github.com/zink-lang/zink/tree/main/evm/opcodes)
  1.1. create a new file `zink/evm/opcodes/src/cancun.rs` (cp from shanghai.rs )
  1.2. add `0x5C` and `0x5D` to the mapping
2. introduce transient storage in the storage macro, could be `#[zink::transient_storage(...)]` comparing with the previous `#[zink::storage(..)]`, for the details of the instructions of transient storage, we just need to replace the `SSTORE` to `TSTORE` and `SLOAD` to `TLOAD`, for example, the [value storage](https://github.com/zink-lang/zink/blob/main/zink/src/storage/value.rs#L15)
  2.1. find a way to optimize the logic between `zink::storage` and `zink::transient_storage` since their logic are the same
3. Provide tests for the transient storage

### PR DESCRIPTION
**Pull request by malik672**
Resolves #260

0x30bE4D758d86cfb1Ae74Ae698f2CF4BA7dC8d693

### Changes

- [x] Support transient storage introduced in Cancun, TSTORE (0x5D) and TLOAD (0x5D)
- [x] add test for transient storage
- [x] example for transient storage

### CODE REVIEW

**Comment by malik672** [comment]
@clearloop

File: examples/transient_storage.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    Ok(())
+}
+
+// Example with both storage types
+#[test]
+fn mixed_storage() -> anyhow::Result<()> {
+    use zint::{Bytes32, Contract, U256};
+
+    let mut contract = Contract::search("mixed_storage")?.compile()?;
```
Comment (clearloop) [inline examples/transient_storage.rs:69]: need to create a new example file for this, 

the `Contract::search("mixed_storage")` will search `examples/mixed_storage` however it does not exist

File: tests/transient_storage.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
+// //! storage tests
```
Comment (clearloop) [inline tests/transient_storage.rs:1]: if you want to test WASM / opcode related stuffs directly,  you need at least fill the logic of  `compiler/filetests/wat/transient_storage/basic.wat` then work based on it

File: zink/src/ffi/evm.rs [hunk 3]
```diff
@@ -11,96 +11,8 @@ extern "C" {
...
-    /// Push 29 bytes to the stack.
-    pub fn push29(val: i32);
-
-    /// Push 30 bytes to the stack.
-    pub fn push30(val: i32);
-
-    /// Push 31 bytes to the stack.
-    pub fn push31(val: i32);
-
+    // ... existing push functions ...
```
Comment (clearloop) [inline zink/src/ffi/evm.rs:14]: ?

File: zink/codegen/src/transient_storage.rs [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    parse::{Parse, ParseStream, Result},
+    parse_quote, Attribute, Ident, ItemFn, ItemStruct, Visibility,
+};
+
+thread_local! {
+   static TRANSIENT_STORAGE_REGISTRY: RefCell<HashSet<String>> = RefCell::new(HashSet::new());
+}
+
+/// Transient storage attributes parser
+pub struct TransientStorage {
```
Comment (clearloop) [inline zink/codegen/src/transient_storage.rs:18]: need to merge this with the exist Storage parser since the logic are the same

**Review by clearloop — CHANGES_REQUESTED** [review]
your approach is correct! however this PR is incomplete, please request for review at least after passing the CI, on the other hand, feel free to request for review if you want to identify if you are in the right direction ( the answer is yes for the current PR! )

1. CI is not passed
2. need to merge the transient storage with storage in the token parser in codegen ( add a new field to Storage for identify if the passed struct is transient or not )
3. missing handlers in the host function registry https://github.com/zink-lang/zink/blob/main/codegen/src/wasm/host.rs

File: tests/transient_storage.rs (hunk 2 above)
Comment (malik672) [reply tests/transient_storage.rs:1]: yeahh but i don't really understand the wat

File: zink/src/ffi/evm.rs (hunk 3 above)
Comment (malik672) [reply zink/src/ffi/evm.rs:14]: lmaooo

**Comment by malik672** [comment]
@clearloop

**Review by clearloop — CHANGES_REQUESTED** [review]
Sorry for the late! Everything looks good now! Please fix the CI 🙏

In case you don't know how to process tests locally

```
# test all tests
cargo tt 

# build all examples
cargo be

# test all examples
cargo te

# run clippy to all packages
cargo cc
```

File: examples/transient_storage.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
+//! Transient Storage example.
+#![cfg_attr(target_arch = "wasm32", no_std)]
+#![cfg_attr(target_arch = "wasm32", no_main)]
+
+extern crate zink;
+
+use zink::storage::TransientStorage;
+
+/// Temporary counter with value type `i32` that resets after each transaction
+#[zink::transient_storage(i32)]
```
[the remark this answers is not in the document]
Comment (malik672) [reply examples/transient_storage.rs:10]: wait, i do not understand

[the remark this answers is not in the document]
Comment (clearloop) [reply examples/transient_storage.rs:10]: please check the [error message](https://github.com/zink-lang/zink/actions/runs/12104019553/job/33746930289?pr=292#step:7:18) in CI,

just need to re-export it here

https://github.com/zink-lang/zink/blob/main/zink/src/lib.rs#L16

**Comment by malik672** [comment]
@clearloop i'm missing something with the exports

**Review by clearloop — COMMENTED** [review]
we don't need `HostFunc::TransientOp` since it is actually `HostFunc::Evm` now ^ ^

**Comment by malik672** [comment]
> we don't need `HostFunc::TransientOp` since it is actually `HostFunc::Evm` now ^ ^

yeah true but the inability of the example to not find the transient macro is suprising

**Comment by malik672** [comment]
@clearloop i think i need help here

File: zink/src/lib.rs [hunk 6]
```diff
@@ -10,10 +10,10 @@ mod event;
 pub mod ffi;
 pub mod primitives;
 pub mod storage;
-
 pub use self::{asm::Asm, event::Event};
-pub use storage::{DoubleKeyMapping, Mapping, Storage};
-pub use zink_codegen::{assert, external, revert, storage, Event};
+pub use storage::{DoubleKeyMapping, Mapping, Storage, TransientStorage};
+pub use zink_codegen::{external, revert, storage, Event};
+pub use zink_codegen::transient_storage;
```
Comment (clearloop) [inline zink/src/lib.rs:16]:
```suggestion
pub use zink_codegen::{assert, external, revert, storage, transient_storage, Event};
```

File: zink/codegen/src/storage.rs [hunk 7]
```diff
@@ -45,14 +91,19 @@ impl Storage {
...
         let is = &self.target;
         let name = self.target.ident.clone();
-        let slot = storage_slot(name.to_string());
+        let slot = self.get_storage_slot(name.to_string());
         let key = slot.to_bytes32();
 
         let keyl = Literal::byte_string(&key);
+        let trait_path = match self.kind {
+            StorageKind::Persistent => quote!(zink::storage::Storage),
+            StorageKind::Transient => quote!(zink::transient_storage::TransientStorage),
```
Comment (clearloop) [inline zink/codegen/src/storage.rs:100]:
```suggestion
            StorageKind::Transient => quote!(zink::storage::TransientStorage),
```

**Review by clearloop — COMMENTED** [review]
Just import-export issues ))

```
error[E0433]: failed to resolve: could not find `transient_storage` in `zink`
  --> examples/transient_storage.rs:10:1
   |
10 | #[zink::transient_storage(i32)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | could not find `transient_storage` in `zink`
   | help: a trait with a similar name exists: `TransientStorage`
   |
   = note: this error originates in the attribute macro `zink::transient_storage` (in Nightly builds, run with -Z macro-backtrace for more info)
```

this is not pointing to the proc-macro but the path `zink::transient_storage::TransientStorage` in your `quote` haha

**Comment by malik672** [comment]
@clearloop transient storage is not present yet

File: examples/transient_storage.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#[zink::external]
+pub fn set_temp(value: i32) {
+    TempCounter::set(value);
+}
+
+/// Get value from the transient storage.
+#[zink::external]
+pub fn get_temp() -> i32 {
+    TempCounter::get()
+}
```
Comment (clearloop) [inline examples/transient_storage.rs:22]:
```suggestion
/// Set and get value via the transient storage.
#[zink::external]
pub fn set_and_get_temp(value: i32) -> i32 {
    TempCounter::set(value);
    TempCounter::get()
}
```

**Review by clearloop — COMMENTED** [review]
> @clearloop transient storage is not present yet in revm cachedb

Oh I saw it, because it is temp storage, in this case, we just need a set_get call to verify our implementation

File: evm/opcodes/src/cancun.rs [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    (0x37, CALLDATACOPY, 3, 3, 0, "Copy input data in current environment to memory.", Frontier, EnvironmentalInformation),
+    (0x38, CODESIZE, 2, 0, 1, "Get size of code running in current environment.", Frontier, EnvironmentalInformation),
+    (0x39, CODECOPY, 3, 3, 0, "Copy code running in current environment to memory.", Frontier, EnvironmentalInformation),
+    (0x3a, GASPRICE, 2, 0, 1, "Get price of gas in current environment", Frontier, EnvironmentalInformation),
+    (0x3b, EXTCODESIZE, 20, 1, 1, "Get size of an account's code.", Frontier, EnvironmentalInformation),
+    (0x3c, EXTCODECOPY, 20, 4, 0, "Copy an account's code to memory.", Frontier, EnvironmentalInformation),
+    (0x3d, RETURNDATASIZE, 2, 0, 1, "Get size of output data from the previous call from the current environment.", Byzantium, EnvironmentalInformation),
+    (0x3e, RETURNDATACOPY, 3, 3, 0, "Copy output data from the previous call to memory.", Byzantium, EnvironmentalInformation),
+    (0x3f, TSTORE, 100, 2, 0, "Store word in transient storage.", Cancun, EnvironmentalInformation),
+    (0x40, TLOAD, 100, 1, 1, "Load word from transient storage.", Cancun, EnvironmentalInformation),
```
Comment (clearloop) [inline evm/opcodes/src/cancun.rs:50]:
```suggestion
    (0x3f, EXTCODEHASH, 100, 1, 1, "Get hash of an account’s code.", Cancun, EnvironmentalInformation),
    (0x40, BLOCKHASH, 20, 1, 1, "Get the hash of one of the 256 most recent complete blocks.", Cancun, EnvironmentalInformation),
```

File: evm/opcodes/src/cancun.rs [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    (0x52, MSTORE, 3, 2, 0, "Save word to memory.", Frontier, StackMemoryStorageFlow),
+    (0x53, MSTORE8, 3, 2, 0, "Save byte to memory.", Frontier, StackMemoryStorageFlow),
+    (0x54, SLOAD, 50, 1, 1, "Load word from storage.", Frontier, StackMemoryStorageFlow),
+    (0x55, SSTORE, 0, 2, 0, "Save word to storage.", Frontier, StackMemoryStorageFlow),
+    (0x56, JUMP, 8, 1, 0, "Alter the program counter.", Frontier, StackMemoryStorageFlow),
+    (0x57, JUMPI, 10, 2, 0, "Conditionally alter the program counter.", Frontier, StackMemoryStorageFlow),
+    (0x58, PC, 2, 0, 1, "Get the value of the program counter prior to the increment.", Frontier, StackMemoryStorageFlow),
+    (0x59, MSIZE, 2, 0, 1, "Get the size of active memory in bytes.", Frontier, StackMemoryStorageFlow),
+    (0x5a, GAS, 2, 0, 1, "Get the amount of available gas.", Frontier, StackMemoryStorageFlow),
+    (0x5b, JUMPDEST, 1, 0, 0, "Mark a valid destination for jumps.", Frontier, StackMemoryStorageFlow),
```
Comment (clearloop) [inline evm/opcodes/src/cancun.rs:72]:
```suggestion
    (0x5b, JUMPDEST, 1, 0, 0, "Mark a valid destination for jumps.", Frontier, StackMemoryStorageFlow),|
   (0x5c, TLOAD, ...),
   (0x5d, TSTORE, ...),
   (0x5e, MCOPY, ...)
```

**Review by clearloop — CHANGES_REQUESTED** [review]
You mapped the wrong opcodes for transient storage, please don't trust AI's output of the opcodes, comparing them us [evm.codes](https://evm.codes) please

**Comment by malik672** [comment]
@clearloop done

**Review by clearloop — APPROVED** [review]
LGTM!  Bounty transaction: https://polygonscan.com/tx/0x53ba55954ef3f1410f61af86885231eef428cd140bab80ad384bacf2ae305541

**Pull request merged**